python/python/tests/test_pubsub.py

```diff
@@ -0,0 +1,1800 @@
+        protocol=protocol,
+    )
+    listening_client = await create_client(
+        request,
+        cluster_mode=cluster_mode,
+        cluster_mode_pubsub=c_pubsub,
+        standalone_mode_pubsub=s_pubsub,
+        protocol=protocol,
+    )
+    return publishing_client, listening_client
+
+
+async def get_message_by_method(
+    method: MethodTesting,
+    client: TRedisClient,
+    messages: Optional[
+        List[Union[RedisClient.PubSubMsg, RedisClusterClient.PubSubMsg]]
+    ] = None,
+    index: Optional[int] = None,
+):
+    if method == MethodTesting.Async:
+        return await client.get_pubsub_message()
+    elif method == MethodTesting.Sync:
+        return client.try_get_pubsub_message()
+    assert messages and (index is not None)
+    return messages[index]
+
+
+async def check_no_messages_left(
+    method,
+    client: TRedisClient,
+):
+    if method == MethodTesting.Async:
+        # assert there are no messages to read
+        with pytest.raises(asyncio.TimeoutError):
+            await asyncio.wait_for(client.get_pubsub_message(), timeout=3)
+    elif method == MethodTesting.Sync:
+        assert client.try_get_pubsub_message() is None
+        # check callback
+
+
+def create_channels_and_patterns(
+    cluster_mode, modes_and_channels: Dict[PubSubChannelModes, Set[str]]
+):
+    result: Dict[
+        Union[
+            RedisClientConfiguration.PubSubChannelModes,
+            ClusterClientConfiguration.PubSubChannelModes,
+        ],
+        Set[str],
+    ] = {}
+    for mode, channel in modes_and_channels.items():
+        if mode == PubSubChannelModes.Exact:
+            if cluster_mode:
+                result[ClusterClientConfiguration.PubSubChannelModes.Exact] = channel
+            else:
+                result[RedisClientConfiguration.PubSubChannelModes.Exact] = channel
+        elif mode == PubSubChannelModes.Pattern:
+            if cluster_mode:
+                result[ClusterClientConfiguration.PubSubChannelModes.Pattern] = channel
+            else:
+                result[RedisClientConfiguration.PubSubChannelModes.Pattern] = channel
+        else:
+            result[ClusterClientConfiguration.PubSubChannelModes.Sharded] = channel
+    return result
+
+
+def create_pubsub_subscription(
+    cluster_mode,
+    channels_and_patterns: Dict[PubSubChannelModes, Set[str]],
+    callback=None,
+    context=None,
+):
+    channels = create_channels_and_patterns(cluster_mode, channels_and_patterns)
+    if cluster_mode:
+        return ClusterClientConfiguration.PubSubSubscriptions(
+            channels_and_patterns=channels,
+            callback=callback,
+            context=context,
+        )
+    return RedisClientConfiguration.PubSubSubscriptions(
+        channels_and_patterns=channels, callback=callback, context=context
+    )
+
+
+def new_message(
+    msg: Union[RedisClient.PubSubMsg, RedisClusterClient.PubSubMsg], context: Any
+):
+    received_messages: List[RedisClient.PubSubMsg] = context
```

Comment:
You're working with this collection from different threads. Are you sure that you don't need synchronisation or locks?

Reply:
why different threads?

Reply:
As I understand, the callback called from another thread, from client's thread pool